src/_utils/datetime-utils.js

```diff
@@ -0,0 +1,149 @@
+const digitRegex = /^[0-9]$/;
+const formatSpecifierRegex = /%[HMdmyY%]/g;
+const daysInWeek = 7;
+
+export function parseDateTime(string, format, _default) {
+  const century = Math.floor((new Date()).getFullYear() / 100);
+
+  if (string === '') {
+    return null;
+  }
+
+  let parts = format.split(formatSpecifierRegex);
+  let stringIdx = 0;
+  let formatIdx = 0;
+  const result = new Date(0);
+
+  for (let part of parts) {
+    if (!string.startsWith(part, stringIdx)) {
+      return _default;
+    }
+
+    stringIdx += part.length;
+    if (stringIdx === string.length) {
+      if (formatIdx !== format.length) {
+        return _default;
+      }
+      break;
+    }
+
+    formatIdx += part.length + 1;
+
+    if (format[formatIdx] === '%') {
+      if (string[stringIdx++] !== '%') {
+        return _default;
+      }
+    } else {
+      let numberLength = 0;
+      while (digitRegex.test(string[stringIdx + numberLength])) {
+        numberLength++;
+      }
+      if (numberLength === 0) {
+        return _default;
+      }
+
+      switch (format[formatIdx]) {
+      case 'H':
+        result.setHours(string.substr(stringIdx, numberLength));
+        break;
+      case 'M':
+        result.setMinutes(string.substr(stringIdx, numberLength));
+        break;
+      case 'd':
+        result.setDate(string.substr(stringIdx, numberLength));
+        break;
+      case 'm':
+        result.setMonth(string.substr(stringIdx, numberLength) - 1);
+        break;
+      case 'y':
+        result.setFullYear(century * 100 + +string.substr(stringIdx, numberLength));
+        break;
+      case 'Y':
+        result.setFullYear(string.substr(stringIdx, numberLength));
+        break;
+      }
```

Comment:
How about putting this in an object outside the function in the form of:
```js
const formats = {
  "H": (result, string) => result.setHours(string),
  // ...
}
```
and call it like:
```js
const substr = string.substr(stringIdx, numberLength);
formats[format[formatIdx]](result, substr);
```
Not exactly, perhaps, but something along these lines

Reply:
Ummm, no... nothing beats `switch` in clarity